src/telemetry.js

```diff
@@ -0,0 +1,132 @@
+/**
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+*/
+
+/* jshint node:true, bitwise:true, undef:true, trailing:true, quotmark:true,
+          indent:4, unused:vars, latedef:nofunc,
+          laxcomma:true
+*/
+
+
+
+// For further details on telemetry, see:
+// https://github.com/cordova/cordova-discuss/pull/43
+
+var Q = require('q');
+
+// Google Analytics tracking code
+var GA_TRACKING_CODE = 'UA-64283057-7';
+
+var pkg = require('../package.json');
+var Insight = require('insight');
+var insight = new Insight({
+    trackingCode: GA_TRACKING_CODE,
+    pkg: pkg
+});
+
+/**
+ * Returns true if the user opted in, and false otherwise
+ */
+function showPrompt() {
+
+    var deferred = Q.defer();
+    
+    var msg = "May Cordova anonymously report usage statistics to improve the tool over time?";
+    insight.askPermission(msg, function (unused, optIn) {
+        var EOL = require('os').EOL;
+        if (optIn) {
+            console.log(EOL + "Thanks for opting into telemetry to help us better cordova");
+        } else {
+            console.log(EOL + "You have been opted out of telemetry. To change this, run: cordova telemetry on");
+            // Always track telemetry opt-outs! (whether opted-in or opted-out)
+            track('telemetry-opt-out', 'via-cli-prompt-choice');
+        }
+        
+        deferred.resolve(optIn); 
+    });
+    
+    return deferred.promise;
+}
+
+function track() {
+    insight.track.apply(insight, arguments);
+}
+
+function trackEvent(category, action, label, value) {
+    insight.trackEvent({
+        category: category,
+        action: action,
+        label: label,
+        value: value
+    });
+}
+
+function turnOn() {
+    insight.optOut = false;
+}
+
+function turnOff() {
+    insight.optOut = true;
+}
+
+/**
+ * Clears telemetry setting
+ * Has the same effect as if user never answered the telemetry prompt
+ * Useful for testing purposes
+ */
+function clear() {
+    insight.optOut = undefined;
+}
+
+function isOptedIn() {
+    return !insight.optOut;
+}
+
+/**
+ * Has the user already answered the telemetry prompt? (thereby opting in or out?)
+ */
+function hasUserOptedInOrOut() {
+    return !(insight.optOut === undefined);
```

Comment:
the typeof === discussion is a matter of preference, I personally have never had an issue with 
```if (x) { // do something } ``` so I can check for null at the same time as I typically set js vars to null, but I am a freak.
As long as you are sure that the value will never actually be null, this should be fine; otherwise change it to == for a simple truthy comparison.
There is however an anti-pattern that is not obvious here. A boolean is being used to hold 3 values, no-answer, user-opted-in, user-opted-out.  When I see ```return !(insight.optOut === undefined);``` my immediate thought would be to correct it to ```return (insight.optOut !== undefined )``` which is not exactly the same meaning.
I would prefer 2 bools in this case. Something like: ```if( wasPrompted && isOptedIn ) ...```
... but again, preference.